// vant/lib/image/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports["default"] = void 0;

var _babelHelperVueJsxMergeProps = _interopRequireDefault(require("@vue/babel-helper-vue-jsx-merge-props"));

var _utils = require("../utils");

var _icon = _interopRequireDefault(require("../icon"));

var _use = (0, _utils.use)('image'),
    sfc = _use[0],
    bem = _use[1];

var _default = sfc({
  props: {
    src: String,
    fit: String,
    alt: String,
    lazyLoad: Boolean,
    width: [String, Number],
    height: [String, Number]
  },
  data: function data() {
    return {
      loading: true,
      error: false
    };
  },
  watch: {
    src: function src() {
      this.loading = true;
      this.error = false;
    }
  },
  computed: {
    style: function style() {
      var style = {};

      if ((0, _utils.isDef)(this.width)) {
        style.width = (0, _utils.suffixPx)(this.width);
      }

      if ((0, _utils.isDef)(this.height)) {
        style.height = (0, _utils.suffixPx)(this.height);
      }

      return style;
    }
  },
  methods: {
    onLoad: function onLoad(event) {
      this.loading = false;
      this.$emit('load', event);
    },
    onError: function onError(event) {
      this.error = true;
      this.loading = false;
      this.$emit('error', event);
    },
    onClick: function onClick(event) {
      this.$emit('click', event);
    },
    renderPlaceholder: function renderPlaceholder() {
      var h = this.$createElement;

      if (this.loading) {
        return h("div", {
          "class": bem('loading')
        }, [this.slots('loading') || h(_icon["default"], {
          "attrs": {
            "name": "photo-o",
            "size": "22"
          }
        })]);
      }

      if (this.error) {
        return h("div", {
          "class": bem('error')
        }, [this.slots('error') || h(_icon["default"], {
          "attrs": {
            "name": "warning-o",
            "size": "22"
          }
        })]);
      }
    },
    renderImage: function renderImage() {
      var h = this.$createElement;
      var imgData = {
        "class": bem('img'),
        attrs: {
          alt: this.alt
        },
        style: {
          objectFit: this.fit
        },
        on: {
          load: this.onLoad,
          error: this.onError
        }
      };

      if (this.error) {
        return;
      }

      if (this.lazyLoad) {
        return h("img", (0, _babelHelperVueJsxMergeProps["default"])([{
          "directives": [{
            name: "lazy",
            value: this.src
          }]
        }, imgData]));
      }

      return h("img", (0, _babelHelperVueJsxMergeProps["default"])([{
        "attrs": {
          "src": this.src
        }
      }, imgData]));
    }
  },
  render: function render(h) {
    return h("div", {
      "class": bem(),
      "style": this.style,
      "on": {
        "click": this.onClick
      }
    }, [this.renderImage(), this.renderPlaceholder()]);
  }
});

exports["default"] = _default;
